pyodide-build/pyodide_build/buildall.py

```diff
@@ -496,6 +497,163 @@ def generate_needs_build_set(
     return needs_build
 
 
+class _GraphBuilder:
+    """A class to manage state for build_from_graph.
+
+    build_from_graph has a bunch of moving parts: a threadpool, a Rich Live
+    session to display progress information to the terminal, and a job graph to
+    keep track of. This class keeps track of all this state.
+
+    The only public class is `run`.
+    """
+
+    pkg_map: dict[str, BasePackage]
+    build_args: BuildArgs
+    build_dir: Path
+    needs_build: set[str]
+    build_queue: PriorityQueue[tuple[int, BasePackage]]
+    built_queue: Queue[tuple[BasePackage, BaseException | None]]
+    lock: Lock
+    building_rust_pkg: bool
+    queue_idx: int
+    progress_formatter: ReplProgressFormatter
```

Comment:
As it is not a dataclass anymore, I guess these variables don't need to be defined in a class variable.

Reply:
Isn't this just the normal way to do type annotations for class attributes?

Reply:
I guess not. Thanks.